---
title: "RiboCryptOverview"
date: "`r BiocStyle::doc_date()`"
author: "Michal Swirski & Hakon Tjeldnes"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{RiboCrypt Overview}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8} 
---

# Introduction

Welcome to the `RiboCrypt` package. 
`RiboCrypt` is an R package for interactive visualization in genomics

This vignette will walk you through usage with examples. 

`RibCrypt` currently supports:  

1. Create interactive browser views for NGS tracks (both transcript and genomic coordinates) 
 with/without frames for Ribo-seq / TCP-seq
2. Interactive reports QC and metaplot analysis

# Creating browser window

The simplest way to create a window is to use the syntax of programing found in the package ORFik.
Here we load a test experiment, to show a region with Ribo-seq:

First let's load the libraries needed
```{r eval = TRUE, echo = TRUE, message = FALSE}
library(RiboCrypt) # This package
library(ORFik)     # The backend package for RiboCrypt
```

Now load the experiment and load the coding sequences (cds), we also load the Ribo-seq library (pshifted reads)
```{r eval = TRUE, echo = TRUE, message = FALSE}
 df <- ORFik.template.experiment()
 # Pick only third library
 df <- df[3,]
 outputLibs(df, type = "pshifted")
 cds <- loadRegion(df, "cds")
```
Now find a cds that has a lot of coverage
```{r eval = TRUE, echo = TRUE, message = FALSE}
  overlaps <- countOverlapsW(cds, RFP, weight = "score")
  overlaps
```
We now plot the cds with highest coverage (we extend flanks by 30 bases), in the resulting plot you can zoom
and check coverage.
```{r eval = TRUE, echo = TRUE, message = FALSE}
  if (requireNamespace("BSgenome.Hsapiens.UCSC.hg19")) {
    multiOmicsPlot_ORFikExp(extendLeaders(extendTrailers(cds[which.max(overlaps)], 30), 30), df = df,
                            reference_sequence = BSgenome.Hsapiens.UCSC.hg19::Hsapiens, 
                            frames_type = "columns")
  }
 
```
